src/lib/comms/include/sol-mqtt.h

```diff
@@ -0,0 +1,393 @@
+#define SOL_MQTT_CONFIG_API_VERSION (1)
+    /**
+     * Should always be set to SOL_MQTT_CONFIG_API_VERSION
+     */
+    uint16_t api_version;
+
+    /**
+     * If set, the broker will drop all messages and subscriptions when
+     * the client disconnects. Must be set if no client id is provided.
+     */
+    bool clean_session;
+
+    /**
+     * Time interval between PING messages that should be sent by the
+     * broker to the client.
+     */
+    time_t keepalive;
+
+    /**
+     * NULL terminated string that should be used as client ID. If
+     * not set, clean_session must be set to true.
+     */
+    char *client_id;
+
+    /**
+     * A message that the broker should send when the client disconnects.
+     */
+    struct sol_mqtt_message *will;
+
+    /**
+     * @brief On connect callback
+     *
+     * @param mqtt MQTT Object
+     * @param data User provided data
+     *
+     * @see sol_mqtt_connect
+     *
+     * Callback called when a connect request has been processed
+     */
+    void (*connect) (struct sol_mqtt *mqtt, void *data);
+
+    /**
+     * @brief On disconnect callback
+     *
+     * @param mqtt MQTT Object
+     * @param data User provided data
+     *
+     * @see sol_mqtt_connect
+     *
+     * Callback called when the client has disconnected from the
+     * broker.
+     */
+    void (*disconnect) (struct sol_mqtt *mqtt, void *data);
+
+    /**
+     * @brief On publish callback
+     *
+     * @param mqtt MQTT Object
+     * @param data User provided data
+     *
+     * @see sol_mqtt_connect
+     *
+     * Callback called when a publish request has been processed.
+     */
+    void (*publish) (struct sol_mqtt *mqtt, void *data);
+
+    /**
+     * @brief On message callback
+     *
+     * @param mqtt MQTT Object
+     * @param message Message received from the broker
+     * @param data User provided data
+     *
+     * @see sol_mqtt_connect
+     *
+     * Callback called when a message is received from the broker.
+     * This does not include PING messages, only messages incoming due
+     * to publish request from other clients or the broker itself.
+     *
+     * The memory associated to the message object will be freed after
+     * the callback returns.
+     */
+    void (*message) (struct sol_mqtt *mqtt, const struct sol_mqtt_message *message, void *data);
+
+    /**
+     * @brief On subscribe callback
+     *
+     * @param mqtt MQTT Object
+     * @param data User provided data
+     *
+     * @see sol_mqtt_connect
+     *
+     * Callback called when a subscribe request has been processed.
+     */
+    void (*subscribe) (struct sol_mqtt *mqtt, void *data);
+
+    /**
+     * @brief On unsubscribe callback
+     *
+     * @param mqtt MQTT Object
+     * @param data User provided data
+     *
+     * @see sol_mqtt_connect
+     *
+     * Callback called when a unsubscribe request has been processed.
+     */
+    void (*unsubscribe) (struct sol_mqtt *mqtt, void *data);
```

Comment:
In soletta we're giving user data as the first parameter for callbacks, see `sol-gpio.h`